WalletWasabi/Blockchain/BlockFilters/IndexBuilderService.cs

```diff
@@ -330,36 +336,40 @@ private void BlockNotifier_OnBlock(object? sender, Block e)
 		found = false; // Only build the filter list from when the known hash is found.
 		var filters = new List<FilterModel>();
 
+		ImmutableList<FilterModel> currentIndex;
+
 		lock (IndexLock)
 		{
-			foreach (var filter in Index)
+			currentIndex = Index;
```

Comment:
However, my gut feeling is lock is needed because the order of updating the reference of `Index` does matter. In that case, you can have a property called `IndexSnapshot` that is updated before the end of lock sections (add, reorg). And here just use snapshot without the lock.

Reply:
> In WW code, there are many instances where we omit synchronization for reads of fields/properties/etc. but it works most of the time

Reference assignment is guaranteed to be atomic on all .NET platforms. 

> latest index snapshot we have (not a stale value, which can be returned if there is no lock

I said multiple things but my suggestion was to have a Snapshot and set it INSIDE the lock. I don't know how compiler works but you don't know either. Let me guess: I believe if you have a property and you keep reading it from multiple threads you will get the updated value except when another thread updates it at that very moment. But even in that case, the difference with the lock is that the moment is before or after the lock. If you update at the moment you read you will get a stale value because the writer will wait for the lock to update. So even with a lock, you will have stale values. 

But I think here you went too deep because it is not required at all to guarantee 20 ns update will propagate immediately. The next read will read the correct value and that is perfect. It is the same with or without the lock.

Reply:
I do not want you to change the code. I just want to dismiss this debate permanently, you brought it up too many times and still not gotten relief from that :(

Reply:
> So I guess it's just you.

Alright, I will take that back then. 🙏 

I read through all that you sent and tried to apply the information to our codebase - with less success. I am bringing up a current example from Wasabi - we had a very similar situation with GetStatus. 

https://github.com/zkSNACKs/WalletWasabi/pull/7854

According to what you wrote this code is wrong, because `RoundStates` reference update is not surrounded by a lock.  Is this correct?

Reply:
Before I reply here, first merge this.